Framework/src/CheckRunner.cxx

```diff
@@ -191,66 +192,83 @@ void CheckRunner::init(framework::InitContext&)
     if (dataRef.header != nullptr && dataRef.payload != nullptr) {
-      auto moArray = ctx.inputs().get<TObjArray*>(input.binding.c_str());
-      mLogger << "Device " << mDeviceName
-              << " received " << moArray->GetEntries()
-              << " MonitorObjects from " << input.binding
-              << ENDM;
+
+      // We don't know what we receive, so we test for an array and then try a tobject.
+      // If we received a tobject, it gets encapsulated in the tobjarray.
+      shared_ptr<const TObjArray> array = nullptr;
+      try {
+        array = ctx.inputs().get<TObjArray*>(input.binding.c_str());
+        mLogger << AliceO2::InfoLogger::InfoLogger::Debug << "CheckRunner " << mDeviceName
+                << " received an array with " << array->GetEntries()
+                << " entries from " << input.binding << ENDM;
+      } catch (runtime_error& e) {
+        // we failed to get the TObjArray, let's try a TObject. If it fails it will throw.
+        shared_ptr<const TObject> tobj = ctx.inputs().get<TObject*>(input.binding.c_str());
+      }
 
       // Check if this CheckRunner stores this input
       bool store = mInputStoreSet.count(DataSpecUtils::label(input)) > 0;
 
-      for (const auto tObject : *moArray) {
+      // for each item of the array, check whether it is a MonitorObject. If not, create one and encapsulate.
+      // Then, store the MonitorObject in the various maps and vectors we will use later.
+      for (const auto tObject : *array) {
         std::shared_ptr<MonitorObject> mo{ dynamic_cast<MonitorObject*>(tObject) };
 
+        if (mo == nullptr) {
+          mLogger << AliceO2::InfoLogger::InfoLogger::Info << "The MO is null, probably a TObject could not be casted into an MO." << ENDM;
+          mLogger << AliceO2::InfoLogger::InfoLogger::Info << "    Creating an ad hoc MO." << ENDM;
+          header::DataOrigin origin = DataSpecUtils::asConcreteOrigin(input);
+          mo = std::make_shared<MonitorObject>(tObject, input.binding, origin.str);
+        }
+
         if (mo) {
-          update(mo);
+          mMonitorObjects[mo->getFullName()] = mo;
+          mMonitorObjectRevision[mo->getFullName()] = mGlobalRevision;
           mTotalNumberObjectsReceived++;
 
           // Add monitor object to store later, after possible beautification
           if (store) {
             mMonitorObjectStoreVector.push_back(mo);
           }
-
-        } else {
-          mLogger << AliceO2::InfoLogger::InfoLogger::Error << "The MO is null, probably a TObject could not be casted into an MO" << ENDM;
         }
       }
     }
   }
+}
 
-  auto qualityObjects = check(mMonitorObjects);
-
-  store(qualityObjects);
-  store(mMonitorObjectStoreVector);
-
-  send(qualityObjects, ctx.outputs());
-
-  // Update global revision number
-  updateRevision();
-
-  // monitoring
-  if (timer.isTimeout()) {
-    timer.reset(1000000); // 10 s.
+void CheckRunner::sendMonitoring()
```

Comment:
This method's name claims that it will send metrics whenever invoked, which is not exactly true, because it checks the timer before doing that. Could you please move checking the timer outside or name it accordingly? Perhaps `sendMonitoringPeriodically`, `sendPeriodicMonitoring` or `trySendingMonitoring`?

Reply:
sendPeriodicMonitoring seems perfect